llms/mlx_lm/models/llama.py

```diff
@@ -30,12 +32,128 @@ def __post_init__(self):
+        if self.rope_type == "llama3":
+            return self.compute_llama3_base_freq()
+        return self.original_base
+
+    def compute_llama3_base_freq(self):
+        factor = self.rope_scaling["factor"]
+        low_freq_factor = self.rope_scaling.get("low_freq_factor", 1.0)
+        high_freq_factor = self.rope_scaling.get("high_freq_factor", 4.0)
+        old_context_len = self.rope_scaling.get(
+            "original_max_position_embeddings", 8192
+        )
 
-            if self.rope_scaling["type"] != "linear":
-                raise ValueError("rope_scaling 'type' currently only supports 'linear'")
+        low_freq_wavelen = old_context_len / low_freq_factor
+        high_freq_wavelen = old_context_len / high_freq_factor
+
+        freqs = self.original_base ** (
+            mx.arange(0, self.dims, 2).astype(mx.float32) / self.dims
+        )
+        new_base_freqs = []
+
+        for freq in freqs:
+            wavelen = 2 * math.pi * freq
+            if wavelen < high_freq_wavelen:
+                new_base_freqs.append(freq)
+            elif wavelen > low_freq_wavelen:
+                new_base_freqs.append(freq * factor)
+            else:
+                assert low_freq_wavelen != high_freq_wavelen
+                smooth = (wavelen - high_freq_wavelen) / (
+                    low_freq_wavelen - high_freq_wavelen
+                )
+                new_base_freqs.append(freq * ((1 - smooth) * factor + smooth))
+
+        return mx.array(new_base_freqs).mean().item()
+
+    def extra_repr(self):
+        return f"{self.dims}, traditional={self.traditional}, max_position_embeddings={self.max_position_embeddings}, scaling_factor={self.scale}, rope_type={self.rope_type}"
+
+    def __call__(self, x, offset: int = 0):
+        seq_len = x.shape[1] + offset
+        base = self.compute_base_freq()
+
+        if seq_len > self.max_position_embeddings:
+            base *= (
+                (self.scale * seq_len / self.max_position_embeddings) - (self.scale - 1)
+            ) ** (self.dims / (self.dims - 2))
+
+        return mx.fast.rope(
+            x,
+            self.dims,
+            traditional=self.traditional,
+            base=base,
+            scale=self.scale,
+            offset=offset,
+        )
+
+
+def initialize_rope(args: ModelArgs):
+    head_dim = args.head_dim or args.hidden_size // args.num_attention_heads
+
+    rope_scaling = args.rope_scaling
+    rope_type = "default"
+    rope_scale = 1.0
+
+    if rope_scaling is not None:
+        scaling_type = rope_scaling.get("type") or rope_scaling.get("rope_type")
+        if scaling_type == "linear":
+            rope_scale = 1 / rope_scaling["factor"]
+            rope_type = "default"
+        elif scaling_type == "llama3":
+            rope_scale = 1.0  # The scaling is handled internally for llama3
+            rope_type = "llama3"
+        else:
+            rope_scale = 2.0  # Default scale if type is not recognized
```

Comment:
Why `2.0` here?

Reply:
Not sure, as far as I can see the default is 1.0 in other places. Should we change to 1.0?

Reply:
I removed this and now it will default to 1.0 as before